introspect/event_manager.go

```diff
@@ -0,0 +1,194 @@
+package introspect
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/ipfs/go-log/v2"
+
+	"github.com/libp2p/go-libp2p-core/event"
+	"github.com/libp2p/go-libp2p-core/introspection/pb"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	"github.com/multiformats/go-multiaddr"
+)
+
+var (
+	jsType     = reflect.TypeOf(new(event.RawJSON)).Elem()
+	peerIdType = reflect.TypeOf(new(peer.ID)).Elem()
+	timeType   = reflect.TypeOf(new(time.Time)).Elem()
+	maddrType  = reflect.TypeOf(new(multiaddr.Multiaddr)).Elem()
+)
+
+type eventManager struct {
+	sync.RWMutex
+	logger *log.ZapEventLogger
+
+	inCh     <-chan interface{}
+	outCh    chan *pb.Event
+	metadata map[reflect.Type]*pb.EventType
+
+	closed  bool
+	closeCh chan struct{}
+	closeWg sync.WaitGroup
+}
+
+func newEventManager(inCh <-chan interface{}) *eventManager {
+	em := &eventManager{
+		inCh:     inCh,
+		logger:   log.Logger("introspection/event-manager"),
+		outCh:    make(chan *pb.Event, cap(inCh)),
+		closeCh:  make(chan struct{}),
+		metadata: make(map[reflect.Type]*pb.EventType),
+	}
+
+	em.closeWg.Add(1)
+	go em.processEvents()
+	return em
+}
+
+func (em *eventManager) Close() error {
+	em.Lock()
+	defer em.Unlock()
+
+	if em.closed {
+		em.closeWg.Wait()
+		return nil
+	}
+
+	close(em.closeCh)
+	em.closeWg.Wait()
+	return nil
```

Comment:
shouldn't we also set `em.closed` to true?

Reply:
Yup. I have a TODO item to tackle closure idempotency. Will take care of that in a follow-up commit before merging.